codice/ddf
# DDF-4447 Patching metacards directly over HTTP should consider the context of sharing attribute changes

#### What does this PR do?
Fixes an issue where direct use of the metacard `PATCH` endpoint could result in sharing permissions being changed incorrectly. 

#### Who is reviewing it? 
@Schachte 
@dimitry-sherman 
@samuelechu 

#### Select relevant component teams: 
@codice/security 

#### Ask 2 committers to review/merge the PR and tag them here.
@garrettfreibott 
@stustison

#### How should this be tested?
Setup the DDF distro with two new users, by modifying `users.properties` and `users.attributes` respectively:

_users.properties_
```
kyle=kyle,group,viewer
steve=steve,group,viewer
```

_users.attributes_
```
"steve" : {
        "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress":"steve@localhost.local"
    },
    "kyle" : {
        "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress":"kyle@localhost.local"
    }
```

Have user `steve` create two search forms and share them with `kyle` at the READ/WRITE level, **not** READ-ONLY or READ/WRITE/SHARE levels (kyle's email should appear on the `access-individuals` attribute if this was done correctly). Use the Karaf CLI or Solr UI to obtain the metacard IDs for these search forms (Solr query: `metacard-tags_txt:query-template`). 

Using the UI, ensure that sharing operations work as expected. `steve` should be able to elevate `kyle`'s privilege to READ/WRITE/SHARE and `kyle` should then be able to share the search form with others. But under no circumstances should `kyle` be able to replace `steve` as owner of the metacard. 

Ensure `kyle` is an `access-individual` on the search form under test. Construct a `PATCH` request to `https://localhost:8993/search/catalog/internal/metacards`, using Postman or another tool, with the following body and **verify failure**:
```
[
    {
        "ids": [
            "2029d29ebe66432d9abc8c0e7e8ad3ce"
        ],
        "attributes": [
            {
                "attribute": "security.access-individuals",
                "values": [
                ]
            },
            {
                "attribute": "security.access-individuals-read",
                "values": [
                ]
            },
            {
                "attribute": "security.access-groups",
                "values": [
                ]
            },
            {
                "attribute": "security.access-groups-read",
                "values": [
                ]
            },
            {
                "attribute": "security.access-administrators",
                "values": [
                	"kyle@localhost.local"
                ]
            },
            {
            	"attribute": "metacard.owner",
                "values": [
                	"steve@localhost.local"
                ]
            }
        ]
    }
]
```

Ensure `kyle` is an `access-individual` on the search form under test. Construct a `PATCH` request to `https://localhost:8993/search/catalog/internal/metacards`, using Postman or another tool, with the following body and **verify failure**:
```
[
    {
        "ids": [
            "2029d29ebe66432d9abc8c0e7e8ad3ce"
        ],
        "attributes": [
            {
                "attribute": "security.access-individuals",
                "values": [
                ]
            },
            {
                "attribute": "security.access-individuals-read",
                "values": [
                ]
            },
            {
                "attribute": "security.access-groups",
                "values": [
                ]
            },
            {
                "attribute": "security.access-groups-read",
                "values": [
                ]
            },
            {
                "attribute": "security.access-administrators",
                "values": [
                ]
            },
            {
            	"attribute": "metacard.owner",
                "values": [
                	"kyle@localhost.local"
                ]
            }
        ]
    }
]
```

#### Any background context you want to provide?
This is only an issue due to the new security attributes that were added on master: 
https://github.com/codice/ddf/pull/3655
Other exposed usages of the catalog framework's update capabilities will be analyzed and tested to ensure no other defects exist. 

#### What are the relevant tickets?
[DDF-4447](https://codice.atlassian.net/browse/DDF-4447)
[DDF-4034](https://codice.atlassian.net/browse/DDF-4034)

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
